Adding an AbstractConv3d interface

gvtulder opened the pull request:
This is an attempt to add an AbstractConv3d interface, similar to the recently added AbstractConv2d but in a 3d version. It's based on and reuses much of the 2d code. There are no optimisations for GPU yet, so it is not very fast.

Would this be a good idea? The code certainly needs polishing, so comments are more than welcome.

nouiz commented:
We discussed this internaly while doing the abstract conv2 and doing the same for 3d was on the list of thing to do. Just not a priority. Thanks for working on that.

I'll let @lamblin and @ballasn comment on the code. They know it more then me. I like the idea of having a common class BaseAbstractConv for the 2d and 3d code. I'm not sure if we should have the same classes for 2d or 3d. I think it is safer to don't share them as you did. But this probably give much code duplicate?

nouiz commented:
Don't push too frequently (even force push). Each time it trigger travis and as it take 2h to run, we can't have too much of it. I killed all previous runs. Mostly, as a travis run takes 2h, try to don't push more frequently then every 2h.

thanks

gvtulder commented:
I see. I had the impression that pushing would cancel the previous build. Sorry.

I wasn't sure about combining the classes for 2d and 3d. At first, I had copied all of the 2d code and updated it for 2d, but that produced a lot of very similar/duplicate code (just with 3 instead of 2 in a few important places). Having separate classes for `AbstractConv2d`, `AbstractConv3d` etc. that share the same base classes seemed like a reasonable solution. But yes, it exposes those `AbstractConv` classes with an ugly `convdim` parameter to indicate 2d/3d.

gvtulder commented:
Rebased and combined a number of small commits. It should still work.

abergeron commented on theano/sandbox/cuda/blas.py:
```diff
@@ -1734,19 +1780,28 @@ class GpuCorr3dMM(BaseGpuCorr3dMM):
         Currently supports "valid" only; "full" can be simulated by setting
         `pad="full"` (at the cost of performance), or by using
         `GpuCorrMM_gradInputs`.
+        The width of a border of implicit zeros to pad the
```
The part you added and the previous part seem to conflict.

abergeron commented on theano/sandbox/cuda/blas.py:
```diff
@@ -1765,8 +1820,10 @@ class GpuCorr3dMM(BaseGpuCorr3dMM):
-    def __init__(self, border_mode="valid", subsample=(1, 1, 1), pad=(0, 0, 0)):
-        super(GpuCorr3dMM, self).__init__(border_mode, subsample, pad)
+    def __init__(self, border_mode="valid", subsample=(1, 1, 1),
+                 filter_dilation=(1, 1, 1), pad=(0, 0, 0)):
```
You should probably have pad default to None, to be able to check that the value is not set through it, like it is done for corrmm 2d

abergeron commented on theano/sandbox/cuda/blas.py:
```diff
@@ -1396,29 +1396,64 @@ class BaseGpuCorr3dMM(GpuOp):
                  pad=(0, 0, 0)):
-        if border_mode != "valid":
-            raise ValueError("border_mode must be 'valid'")
+        if pad != (0, 0, 0):
```
Have pad default to None and check that value here.

abergeron commented on theano/tensor/nnet/__init__.py:
```diff
@@ -150,3 +151,120 @@ def conv2d(input, filters, input_shape=None, filter_shape=None,
+
+def conv3d(input, filters, input_shape=None, filter_shape=None,
+           border_mode='valid', subsample=(1, 1, 1), filter_flip=True,
+           image_shape=None, filter_dilation=(1, 1, 1), **kwargs):
```
Since you don't need to provide backwards compatibility because this is a new function you can just import conv3d and not have a wrapper function.  The checks in conv2d are only there because there was another conv2d function that got replaced.

You should probably move the documentation though while dropping references to deprecated things.

abergeron commented on theano/tensor/nnet/abstract_conv.py:
```diff
@@ -657,34 +946,46 @@ class BaseAbstractConv2d(Op):
+            if len(border_mode) != convdim:
+                raise ValueError(
+                    'border mode must have exactly {} values, '
+                    'but was {}'.format(convdim))
```
The format string for the error expects two values, but you provide only one.

abergeron commented:
jenkins test this please

gvtulder commented:
@abergeron: Thanks for the comments. I fixed the things you mentioned.

I now also added a patch to add nnet.conv3d to the library/nnet/conv documentation.

nouiz commented:
jenkins test this

gvtulder commented on doc/library/tensor/nnet/conv.txt:
```diff
@@ -31,13 +31,18 @@
     that will be replaced by an actual convolution implementation during
     the optimization phase.
 
+    As of October 2016, there is also a conv3d interface that provides a similar
```
Done.

gvtulder commented:
The Jenkins test failed because I had missed one `pad=None` in the old gpu backend. That should be fixed now.

ballasn commented on theano/sandbox/cuda/opt.py:
```diff
@@ -2737,7 +2740,10 @@ def local_inplace_gpu_sparse_block_outer(node):
+                  AbstractConv3d,
+                  AbstractConv3d_gradWeights,
+                  AbstractConv3d_gradInputs])
 def local_conv2d_gpu_conv(node):
```
We might want to change  the name if we use this optimization also for conv3d

gvtulder replied:
Changed.

lamblin commented on doc/library/tensor/nnet/conv.txt:
```diff
@@ -190,12 +195,31 @@ TODO: Give examples on how to use these things! They are pretty complicated.
+
+    - :func:`dnn_conv3d <theano.sandbox.cuda.dnn.dnn_conv3d>` GPU-only
+      convolution using NVIDIA's cuDNN library. This requires that you have
+      cuDNN installed and available, which in turn requires CUDA 6.5 and a GPU
```
I think we only support cuDNN v4 or more in the master, which requires CUDA 7.0 or later.

gvtulder replied:
I've replaced this for dnn_conv3d and dnn_conv2d.

lamblin commented on doc/library/tensor/nnet/conv.txt:
```diff
@@ -190,12 +195,31 @@ TODO: Give examples on how to use these things! They are pretty complicated.
+      If cuDNN is available, by default, Theano will replace all nnet.conv3d
+      operations with dnn_conv3d. To explicitly disable it, set
+      ``THEANO_FLAGS=optimizer_excluding=conv_dnn`` in your environment.
+      As dnn_conv3d has a gradient defined, you can also use it manually.
```
maybe ``dnn_conv3d``

gvtulder replied:
I'm not entirely sure what you mean by that.

lamblin replied:
Sorry, maybe double back-ticks around "dnn_conv3d" for formatting in a monospace font is what I meant.

gvtulder replied:
I see. I started to edit that, but then got confused by all the other things that aren't between back-ticks. For example, shouldn't nnet.conv3d also be monospace? There are quite a few examples elsewhere in this section as well.

lamblin replied:
It's not important, and indeed it is not respected everywhere, I just pointed out one that I noticed.
"nnet.conv3d" could indeed be monospace, either because of back-ticks or via a `:func:` link.
If the PR is otherwise ready, don't bother with that.

gravity1989 commented:
Is this a part of the branch? Can't seem to find it. This is very useful for me.

Is there a way I can use this?

gvtulder commented:
@gravity1989: It's not yet included in the Theano master branch, but you can find the code in the `f-abstractconv3d` branch in my repository: https://github.com/gvtulder/Theano/tree/f-abstractconv3d

Everything should work, so it would be nice if you could give it a try. You can either check out that branch (it doesn't include the Theano patches from the last two weeks) or merge it into your local repository.

gravity1989 commented:
I've never tried building theano myself.

Guess have only used pip install. Is it easy?

On Thu, Oct 6, 2016 at 3:45 PM, Gijs van Tulder <notifications@github.com>
wrote:

> @gravity1989 <https://github.com/gravity1989>: It's not yet included in
> the Theano master branch, but you can find the code in the
> f-abstractconv3d branch in my repository: https://github.com/gvtulder/
> Theano/tree/f-abstractconv3d
>
> Everything should work, so it would be nice if you could give it a try.
> You can either check out that branch (it doesn't include the Theano patches
> from the last two weeks) or merge it into your local repository.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Theano/Theano/pull/4862#issuecomment-251921609>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABkKwJzU4U24uuX9yaOnkTqCkUEej7Dxks5qxMongaJpZM4Jk4wJ>
> .
>



-- 
Regards,
R.V.

gvtulder commented:
@gravity1989: It's not difficult. There are a few examples in the [manual](http://deeplearning.net/software/theano/install.html#bleeding-edge-install-instructions). You can even use pip to install. If you just want to try the new code without installing, you can also `git clone` the repository and put the `theano/` directory in your path or in your project's directory, where Python can find it.
However, this pull request probably isn't the best place for this type discussion, so you should probably send your installation questions to the mailing list.

gvtulder commented:
I've now added this to the new gpu backend as well. There was no `dnn_conv3d` yet, so I added those 3D methods and reused the 3D tests from the old backend.

It works, but only if you run it with `dnn.conv.algo_fwd=none`. The default setting for `dnn.conv.algo_fwd` is small, which doesn't seem to work for 3D convolution. I'm not sure how best to solve this. I tried to add algo='none' explicitly in `dnn_conv3d`, but then it still breaks if one of the `grad` methods introduces a new convolution with the default algorithm.

gvtulder commented:
Rebased to the current master.

nouiz commented:
Sadly it still need a rebase. @abergeron is good with the PR @ballasn should finish the check today. So we should be able to merge in the next few days.

thanks

gvtulder commented:
Rebased again. (The conflict was a small change in the comments of AbstractConv's flops method.)

abergeron commented:
jenkins test this please

ballasn commented:
THe PR looks good to me, thanks for the good work!

gvtulder commented:
Good. Thanks for reviewing.

slefrancois commented:
jenkins test this please
